File: apps/cms/src/collections/SaveeUsers.ts
import { CollectionConfig } from "payload";

export const SaveeUsers: CollectionConfig = {
  slug: "savee_users",
  labels: {
    singular: "Savee User",
    plural: "Savee Users",
  },
  admin: {
    description: "Savee.com user profiles discovered during scraping",
    useAsTitle: "display_name",
    defaultColumns: [
      "avatar",
      "username",
      "display_name",
      "follower_count",
      "saves_count",
      "last_scraped_at",
      "is_active",
    ],
    group: "Savee Data",
  },
  access: {
    read: () => true,
    create: () => true,
    update: () => true,
    delete: () => true,
  },
  fields: [
    // Avatar preview (UI only)
    {
      name: "avatar",
      type: "ui",
      admin: {
        components: {
          Cell: "@/components/SaveeUserAvatarCell",
        },
      },
      label: "Avatar",
    },
    // Basic profile info
    {
      name: "username",
      type: "text",
      required: true,
      unique: true,
      admin: {
        description: "Savee.com username (unique identifier)",
      },
    },
    {
      name: "display_name",
      type: "text",
      admin: {
        description: "Display name shown on profile",
      },
    },
    {
      name: "bio",
      type: "textarea",
      admin: {
        description: "User bio/description",
      },
    },
    {
      name: "profile_image_url",
      type: "text",
      admin: {
        description: "URL to profile avatar image",
      },
    },
    {
      name: "avatar_r2_key",
      type: "text",
      admin: {
        description: "R2 key for stored avatar (when available)",
      },
    },
    {
      name: "cover_image_url",
      type: "text",
      admin: {
        description: "URL to profile cover/banner image",
      },
    },
    {
      name: "profile_url",
      type: "text",
      required: true,
      admin: {
        description: "Full URL to Savee profile page",
      },
    },

    // Statistics
    {
      name: "follower_count",
      type: "number",
      defaultValue: 0,
      admin: {
        description: "Number of followers",
      },
    },
    {
      name: "following_count",
      type: "number",
      defaultValue: 0,
      admin: {
        description: "Number of people this user follows",
      },
    },
    {
      name: "saves_count",
      type: "number",
      defaultValue: 0,
      admin: {
        description: "Total number of saves/items",
      },
    },
    {
      name: "collections_count",
      type: "number",
      defaultValue: 0,
      admin: {
        description: "Number of collections created",
      },
    },

    // Additional info
    {
      name: "location",
      type: "text",
      admin: {
        description: "User location (if provided)",
      },
    },
    {
      name: "website_url",
      type: "text",
      admin: {
        description: "Personal website URL",
      },
    },
    {
      name: "social_links",
      type: "json",
      admin: {
        description: "Social media links and profiles",
      },
    },

    // Status
    {
      name: "is_verified",
      type: "checkbox",
      defaultValue: false,
      admin: {
        description: "Whether the user has a verified account",
      },
    },
    {
      name: "is_active",
      type: "checkbox",
      defaultValue: true,
      admin: {
        description: "Whether the account is still active",
      },
    },
    {
      name: "last_scraped_at",
      type: "date",
      admin: {
        description: "When this profile was last updated/scraped",
        date: {
          pickerAppearance: "dayAndTime",
        },
      },
    },
    {
      name: "first_discovered_at",
      type: "date",
      admin: {
        description: "When this user was first discovered",
        date: {
          pickerAppearance: "dayAndTime",
        },
      },
    },
  ],

  // Database table options
  dbName: "savee_users",
};

File: apps/worker/app/models/savee_users.py
"""
Savee Users model for tracking Savee.com user profiles
"""
from sqlalchemy import Column, Integer, String, DateTime, Text, Boolean, JSON
from sqlalchemy.sql import func
from .base import Base


class SaveeUser(Base):
    """Represents a Savee.com user profile"""
    __tablename__ = "savee_users"

    id = Column(Integer, primary_key=True, autoincrement=True)
    username = Column(String(100), unique=True, nullable=False, index=True)
    display_name = Column(String(200), nullable=True)
    bio = Column(Text, nullable=True)
    profile_image_url = Column(String(500), nullable=True)
    avatar_r2_key = Column(String(500), nullable=True)
    cover_image_url = Column(String(500), nullable=True)
    profile_url = Column(String(500), nullable=False)
    
    # Statistics
    follower_count = Column(Integer, default=0)
    following_count = Column(Integer, default=0)
    saves_count = Column(Integer, default=0)
    collections_count = Column(Integer, default=0)
    
    # Profile metadata
    location = Column(String(200), nullable=True)
    website_url = Column(String(500), nullable=True)
    social_links = Column(JSON, nullable=True)  # Store social media links
    
    # Status and tracking
    is_verified = Column(Boolean, default=False)
    is_active = Column(Boolean, default=True)
    last_scraped_at = Column(DateTime(timezone=True), nullable=True)
    first_discovered_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    def __repr__(self):
        return f"<SaveeUser(id={self.id}, username='{self.username}', display_name='{self.display_name}')>"


